### ISSUE TITLE
Can't make cusolver LU decomposition work

### ISSUE THREAD

**Issue by itakatz** [lebedov/scikit-cuda#184]
Hi,
I am running the following code, to preform LU decomposition:
```
import numpy as np
import pycuda.autoinit
import pycuda.gpuarray as gpuarray
import skcuda.cusolver as cusolver

# create a cuSolver context
cu_handle = cusolver.cusolverDnCreate()

a = np.array([8,0,0,0,0,-1,8,0,0,0,0,-1,8,0,0,0,0,1,8,0,0,0,0,-1,8],dtype=np.float32).reshape((5,5))
# copy to gpu
a_gpu = gpuarray.to_gpu(a)
# call the LU factorization method
M, N = a_gpu.shape
lda = M
l_work = cusolver.cusolverDnSgetrf_bufferSize(cu_handle, M, N, int(a_gpu.gpudata), lda)
w_buffer = gpuarray.zeros(l_work, dtype=a_gpu.dtype)
devIpiv = gpuarray.zeros(M, dtype=int)
devInfo = gpuarray.zeros(1, dtype=int)
cusolver.cusolverDnSgetrf(cu_handle, M, N, int(a_gpu.gpudata), lda, int(w_buffer.gpudata), int(devIpiv.gpudata), int(devInfo.gpudata))
# destroy the context
cusolver.cusolverDnDestroy(cu_handle)
```

However, this does not seem to return the correct result. Specifically, `a_gpu` is unchanged (supposed to be overwritten), and `devIpiv` has permutation indices which are clearly wrong:
`[ 8589934593, 17179869187, 5, 0, 0]`

Btw, If I do the same using the C interface to CUDA (by editing the sample code in `cuda_samples/7_CUDALibraries/cuSolverDn_LinearSolver/cuSolverDn_LinearSolver.cpp`), it seems to work correctly.
Any ideas?

Thank you.

**Comment by lebedov** [comment]
Try transposing the matrix or saving it with column-major order before copying it to the GPU; numpy assumes row-major order.

**Comment by itakatz** [comment]
Thanks -
I've tried it, but the result is the same. Don't the values of `devIpiv` suggest that something's wrong?

**Comment by lebedov** [comment]
I should have mentioned that you need to explicitly copy the matrix after transposing (i.e., `a_gpu = gpuarray.to_gpu(a.T.copy())`) because numpy's transpose is actually a view - see `demos/cusolver_demo_lu.py`. In any case, the skcuda function appears to be returning the same value as scipy - try computing `scipy.linalg.lu(a.T)`. Not sure why you are getting weird pivots - I didn't get them when I ran your code.

**Comment by itakatz** [comment]
First, thank you very much for the fast response. It's not to be taken for granted.

After you 1st suggestion, I created `a` with the `order='F'` option, but I did it before the reshape, so I was creating a 1D array in `F` mode and then reshaped it, which is wrong.
Regarding the demos folder - I wasn't aware of it, it could have saved me a lot of time yesterday reading the cuSolver docs. Thanks for pointing it out.

Lastly, regarding the wrong `devIpiv` values, it seemed like a result of wrong data-type, and indeed if I define it using 32 bit int:
`devIpiv = gpuarray.zeros(M, dtype=np.int32)`
It has the correct values after the function call.

Thanks again for your help
Itamar.
